### ISSUE TITLE
Explicit quantum parallelism

### ISSUE THREAD

**Issue by samarsha** [microsoft/qsharp#492]
From microsoft/qsharp-language#162:

> **Is your feature request related to a problem? Please describe.**
> 
> There is a crucial interaction between parallelism and memory management in quantum computing: Depending on the memory management strategy, the compiler (or the run-time system) may introduce artificial dependencies, e.g., among loop iterations that would otherwise be subject to parallelism. As an example, consider the following code snippet:
> 
> ```qsharp
> for (q in qubits) {
>   Op(q);
> }
> ```
> 
> `Op` is some quantum operation that takes a qubit as input. If `Op` is in the native gate set of the target architecture, for example, then this loop may be executed in parallel. However, if `Op` is temporarily allocates and then deallocates a helper qubit, quantum memory management may decide to reuse the same helper qubit for all iterations of the loop.
> 
> ![image](https://user-images.githubusercontent.com/5529874/190448799-4379ee76-8d30-47f6-9a20-be63a3b2e9fa.png)
> 
> An example of this is given in above. On the right implementation of a controlled rotation gate is shown that uses one helper qubit to avoid doubling the number of single-qubit rotations, which induces a large overhead when implemented fault-tolerantly. Consequently, all potential parallelism is lost due to the introduction of this dependency. The no-cloning theorem prohibits removing this dependency through copying. Furthermore, an alternative resource management strategy that opts to never reuse resources would result in prohibitively large numbers of qubits and gates. As a solution to this problem, we propose language constructs to explicitly mark parallel regions.
> 
> **Describe the solution you'd like** We proposed a language extension, with the example above could be expressed as
> 
> ```
> parallel for (q in qubits) {
>   Op(q);
> }
> ```
> 
> where the `parallel` indicates that managed resources should not be reused across loop iterations. In more details, the syntax and semantics of our proposed language extension is provided below. QParallel provides support for parallel sections and loops, in addition to specific clauses that facilitate making existing code parallel.
> 
> _**Parallel sections**_
> 
> A `parallel sections` statement defines a portion of the program that should be run in parallel. The syntax is as follows:
> 
> ```
> parallel sections {
>   SectionSequence
> }
> ```
> 
> The code encapsulated in a `parallel sections` block contains a sequence of section statements. Each section should be run in parallel with other section code blocks, if possible.
> 
> _**Section statement**_
> 
> The section statement marks a section that can be run in parallel with other section blocks in the same `parallel sections` environment.
> 
> ```
> section {
>   StatementBlock
> }
> ```
> 
> A precondition to the usage of this statement is that each section can be executed in parallel. Consequently, different sections must not share any qubits. In particular, to ensure parallel execution, automatic memory management must not reuse temporary qubits that are allocated and deallocated inside a different section of the same parallel sections block.
> 
> _**Parallel for statement**_
> 
> Another case that commonly benefits from parallelization is a loop where each iteration of the loop can be run in parallel. To support this use case, we introduce a `parallel for` statement. Its behavior is identical to that of a `parallel sections` statement, where each iteration of a loop is considered a separate section inside of it. The `parallel for` statement can be used as follows:
> 
> ```
> parallel for i in iterable {
>   StatementBlock
> }
> ```
> 
> _**Fanout clause**_
> 
> Consider the following for-loop in Q#:
> 
> ```
> for t in targets {
>   CNOT(control, t);
> }
> ```
> 
> This loop cannot be executed in parallel because all `CNOT` operations share the same `control` qubit, declared outside of the loop. As a remedy, we introduce the so-called `fanout` clause, which allows us to rewrite the code above as a parallel for-loop with `fanout`:
> 
> ```
> parallel for t in targets fanout(control, 4) {
>   CNOT(control, t);
> }
> ```
> 
> In this example, the control qubit is fanned out to three extra qubits, resulting in a total of 4 control qubits that are shared among loop iterations. This reduces the runtime of the loop by up to 4×.
> 
> More generally, a `fanout(qubits, n)` clause fans out the specified `qubits` to n-1 temporary qubits, so that loop iterations can use one of these "entangled copies" and thus remove some or all of the data dependencies. At the end of the loop, an unfanout operation, which is the inverse of the initial fanout operation, will consolidate the effect of all iterations on their respective copies of the fanned-out qubit(s).
> 
> _**BNF**_
> 
> Backus-Naur form below describes the syntax of our proposed language extension.
> 
> ![image](https://user-images.githubusercontent.com/5529874/190453146-912466db-6f6f-41b4-aa7a-9c9673b4a864.png)
> 
> **Describe alternatives you've considered** Alternative is to add Q# intrinsics to directly call qubit manager functionality implemented in C# and C++ without any changes to the language. It would be also nice to add documentation explaining how this intrinsic can be used and some providing examples.
> 
> C# runtime implements required functionality here:
> 
> * [StartRestrictedReuseArea](https://github.com/microsoft/qsharp-runtime/blob/84f81d6f72f60f36b2136e72a5321c20d6aef75b/src/Simulation/QCTraceSimulator/QubitManagerRestrictedReuse.cs#L165)
> * [NextRestrictedReuseSegment](https://github.com/microsoft/qsharp-runtime/blob/84f81d6f72f60f36b2136e72a5321c20d6aef75b/src/Simulation/QCTraceSimulator/QubitManagerRestrictedReuse.cs#L170)
> * [EndRestrictedReuseArea](https://github.com/microsoft/qsharp-runtime/blob/84f81d6f72f60f36b2136e72a5321c20d6aef75b/src/Simulation/QCTraceSimulator/QubitManagerRestrictedReuse.cs#L182)
> 
> and C++ QIR runtime implements required functionality here:
> 
> * [StartRestrictedReuseArea](https://github.com/microsoft/qsharp-runtime/blob/2eb0f3fd14bf62dd7dcb3cab992cb505539aa43f/src/Qir/Runtime/public/QubitManager.hpp#L48)
> * [NextRestrictedReuseSegment](https://github.com/microsoft/qsharp-runtime/blob/2eb0f3fd14bf62dd7dcb3cab992cb505539aa43f/src/Qir/Runtime/public/QubitManager.hpp#L49)
> * [EndRestrictedReuseArea](https://github.com/microsoft/qsharp-runtime/blob/2eb0f3fd14bf62dd7dcb3cab992cb505539aa43f/src/Qir/Runtime/public/QubitManager.hpp#L50)
> 
> **Additional context** This will allow to avoid issues similar to the ones discussed in: [microsoft/qsharp-runtime#1037](https://github.com/microsoft/qsharp-runtime/issues/1037) [microsoft/qsharp-runtime#419](https://github.com/microsoft/qsharp-runtime/issues/419) and make quantum parallelism more intuitive for developers . Above features can simplify and improve code in https://github.com/microsoft/QuantumEllipticCurves. Detailed analysis of this feature and results obtained using a prototype implementation are available at [arXiv.2210.03680](https://doi.org/10.48550/arXiv.2210.03680).

**Comment by swernli** [comment]
Superseded by #3274 and fixed in #3298
